docs: rewrite introduction to dependency injection (#44466)
angular/angular: aio/content/guide/dependency-injection.md

# Understanding dependency injection

Dependency injection, or DI, is one of the fundamental concepts in Angular. DI is wired into the Angular framework and allows classes with Angular decorators, such as Components, Directives, Pipes, and Injectables, to configure dependencies that they need. 

Two main roles exist in the DI system: dependency consumer and dependency provider. 

> Angular facilitates the interaction between dependency consumers and dependency providers using an abstraction called [Injector](guide/glossary#injector). When a dependency is requested, the Injector checks its registry to see if there is an instance already available there and if not, a new instance is created and stored in the registry. Angular creates an application-wide injector (also known as "root" injector) during the application bootstrap process, as well as any additional injectors as needed. You don't need to create injectors manually, but it's important to know that there is a layer that connects providers and consumers.

Reviewer: Small note on creating injectors manually: for some APIs it's still needed, so we can update the doc to mention that.
```suggestion
Angular facilitates the interaction between dependency consumers and dependency providers using an abstraction called [Injector](guide/glossary#injector). When a dependency is requested, the Injector checks its registry to see if there is an instance already available there and if not, a new instance is created and stored in the registry. Angular creates an application-wide injector (also known as "root" injector) during the application bootstrap process, as well as any additional injectors as needed. You don't need to create injectors manually in majority of use-cases, but it's important to know that there is a layer that connects providers and consumers.
```

Author: I added this to the sentence but to make it concise, I rewrote it to be: In most cases you don't need to manually create injectors, but you should know that there is a layer that connects providers and consumers.

This topic covers basic scenarios of how a class can act as a dependency. However, Angular allows you to use functions, objects, primitive types such as string or Boolean, or any other types as dependencies as well. For more information, see [Dependency providers](guide/dependency-injection-providers].

## Providing dependency

Imagine there is a class called HeroService that needs to act as a dependency in a component.

The first step is to add the @Injectable decorator to indicate that the class can be injected.

<code-example language="sh">
@Injectable()
class HeroService {}
</code-example>

The next step is to make it available in the DI by providing it.  A dependency can be provided in multiple places:

* At the Component level, using the `providers` field of the `@Component` decorator. In this case the `HeroService` becomes available to all instances of this component and other components and directives used in the template. For example:

> <code-example language="sh">

Reviewer: I think we should use "typescript" as a value here (and in other cases that we've updated in this doc):
```suggestion
<code-example language="typescript">
```

Author: Fixed.

validating user input, or logging directly to the console.
By defining such processing tasks in an *injectable service class*, you make those tasks
available to any component.
You can also make your application more adaptable by injecting different providers of the same kind of service,
as appropriate in different circumstances.

Angular doesn't *enforce* these principles. Angular does help you *follow* these principles
by making it easy to factor your application logic into services and make those services
available to components through DI.

### Service examples

Here's an example of a service class that logs to the browser console.

<code-example path="architecture/src/app/logger.service.ts" header="src/app/logger.service.ts (class)" region="class"></code-example>

Services can depend on other services. For example, here's a `HeroService` that depends on the `Logger` service, and also uses `BackendService` to get heroes. That service in turn might depend on the `HttpClient` service to fetch heroes asynchronously from a server.

<code-example path="architecture/src/app/hero.service.ts" header="src/app/hero.service.ts (class)" region="class"></code-example>

> ## What's next

Reviewer: I noticed that the "What's next" section on different pages contain different links (some pages have no "What's next" section). I think it'd be great to have a consistent list of links on each page (the list of all DI topics that we have on the main DI page). In this case, users who came to a particular page can navigate to a different article.

# Understanding Dependency injection

> Dependency injection, or DI, is wired into the Angular framework and used everywhere to provide new components with the services or other things they need.
> Components consume services; that is, you can *inject* a service into a component, giving the component access to that service class.

Reviewer: I'd propose to rephrase this a bit:
```suggestion
Dependency Injection (or DI) is a one of the fundamental concepts in Angular. DI is wired into the framework and allows classes with Angular decorators (such as Components, Directives, Pipes and Injectables) to configure dependencies that they need. You can *inject* a service into a class using its constructor, giving that class an access to a particular dependency.
```

Author: Incorporated.

## Services and dependency injection

*Service* is a broad category encompassing any value, function, or feature that an application needs.
A service is typically a class with a narrow, well-defined purpose.

> It should do something specific and do it well.

Reviewer: I think we can drop that sentence, it doesn't seem to add value:
```suggestion
```

Author: Sentence dropped.

> Angular distinguishes components from services to increase modularity and reusability.

Reviewer: The "component" term is used in the content below. While it makes sense as an example, this may set the wrong impression that components are the only type of classes that can use DI. In fact other classes can do that too. I think it'd be great to reshuffle the content a bit to give a very high-level overview first that gives a definition of what DI pattern is, how Angular enables it and which classes can participate in DI. After that we can say something like "let's used Components as an example and see how DI can be useful" and have a couple paragraphs dedicated to that. We can chat more offline, I can help structure this part of the doc.

Author: I restructured the topic.

# Understanding Dependency injection

Dependency injection, or DI, is wired into the Angular framework and used everywhere to provide new components with the services or other things they need.
Components consume services; that is, you can *inject* a service into a component, giving the component access to that service class.

> ## Services and dependency injection

Author: Used Dependency injection in first instance, and DI in later instances.

Ideally, a component's job is to enable the user experience and nothing more.
A component should present properties and methods for data binding,
in order to mediate between the view (rendered by the template)
and the application logic (which often includes some notion of a *model*).

A component can delegate certain tasks to services, such as fetching data from the server,
validating user input, or logging directly to the console.
By defining such processing tasks in an *injectable service class*, you make those tasks
available to any component.
You can also make your application more adaptable by injecting different providers of the same kind of service,
as appropriate in different circumstances.

Angular doesn't *enforce* these principles. Angular does help you *follow* these principles
by making it easy to factor your application logic into services and make those services
available to components through *dependency injection*.

### Service examples

Here's an example of a service class that logs to the browser console.

<code-example path="architecture/src/app/logger.service.ts" header="src/app/logger.service.ts (class)" region="class"></code-example>

> Services can depend on other services. For example, here's a `HeroService` that depends on the `Logger` service, and also uses `BackendService` to get heroes. That service in turn might depend on the `HttpClient` service to fetch heroes asynchronously from a server.

Reviewer 2: ```suggestion
Services can depend on other services. For example, here's a `HeroService` that depends on the `Logger` service, and also uses `BackendService` to get heroes. That service, in turn, might depend on the `HttpClient` service to fetch heroes asynchronously from a server.
```

<code-example path="architecture/src/app/hero.service.ts" header="src/app/hero.service.ts (class)" region="class"></code-example>

## Creating Dependency injection (DI)

<img src="generated/images/guide/architecture/dependency-injection.png" alt="Service" class="left">

> To define a class as a service in Angular, use the `@Injectable()` decorator to provide the metadata that allows Angular to inject it into a component as a *dependency*.

Reviewer 2: ```suggestion
To define a class as a service in Angular, use the `@Injectable()` decorator to provide the metadata that allow Angular to inject it into a component as a *dependency*.
```

Similarly, use the `@Injectable()` decorator to indicate that a component or other class (such as another service, a pipe, or an NgModule) *has* a dependency.

* The *injector* is the main mechanism. Angular creates an application-wide injector for you during the bootstrap process, and additional injectors as needed. You don't have to create injectors.

> * An injector creates dependencies, and maintains a *container* of dependency instances that it reuses if possible.

Reviewer 2: ```suggestion
* An injector creates dependencies, and maintains a *container* of dependency instances that it reuses, if possible.
```

* A *provider* is an object that tells an injector how to obtain or create a dependency.

> For any dependency that you need in your app, you must register a provider with the application's injector,

Reviewer 2: ```suggestion
For any dependency that you need in your app, you must register a provider with the application's injector
```

so that the injector can use the provider to create new instances.
For a service, the provider is typically the service class itself.

<div class="alert is-helpful">

> A dependency doesn't have to be a service&mdash;it could be a function, for example, or a value.

Reviewer 2: There is a specific way to enter an em dash in Cider; this isn't it.

</div>

When Angular creates a new instance of a component class, it determines which services or other dependencies that component needs by looking at the constructor parameter types. For example, the constructor of `HeroListComponent` needs `HeroService`.

<code-example path="architecture/src/app/hero-list.component.ts" header="src/app/hero-list.component.ts (constructor)" region="ctor"></code-example>

> When Angular discovers that a component depends on a service, it first checks if the injector has any existing instances of that service. If a requested service instance doesn't yet exist, the injector makes one using the registered provider, and adds it to the injector before returning the service to Angular.

Reviewer 2: ```suggestion
When Angular discovers that a component depends on a service, it first checks if the injector has any existing instances of that service. If a requested service instance doesn't yet exist, the injector makes one using the registered provider and adds it to the injector before returning the service to Angular.
```
